cognite/client/data_classes/data_modeling/dsl_filter.py

```diff
@@ -0,0 +1,180 @@
+from __future__ import annotations
+
+from dataclasses import asdict, dataclass
+from typing import Dict, List, Literal, Union, get_args
+
+from cognite.client.data_classes.data_modeling.shared import load_reference
+
+__all__ = [
+    "RawPropertyList",
+    "ReferencePropertyValue",
+    "FilterObject",
+    "FilterPropertyObject",
+    "EqualObject",
+    "InObject",
+    "RangeValue",
+    "RangeObject",
+    "ParameterizedPropertyValue",
+    "PrefixObject",
+    "ExistsObject",
+    "ContainsAnyObject",
+    "ContainsAllObject",
+    "OverlapObject",
+    "Logical",
+    "ComplexOperation",
+    "PropertyOperation",
+    "FilterKey",
+    "DSLFilter",
+    "NestedObject",
+    "load_dsl_filter",
+    "dump_dsl_filter",
+]
+
+RawPropertyList = Union[List[float], List[bool], List[str]]
+RawPropertyValue = Union[str, float, bool, dict, RawPropertyList]
+
+
+@dataclass
+class FilterObject:
+    @classmethod
+    def load(cls, data: dict) -> FilterObject:
+        return cls(**data)
+
+
+@dataclass
+class FilterPropertyObject(FilterObject):
+    property: list[str]
+
+
+@dataclass
+class ReferencePropertyValue:
+    property: list[str]
+
+
+@dataclass
+class EqualObject(FilterPropertyObject):
+    value: RawPropertyValue | ReferencePropertyValue
+
+
+@dataclass
+class InObject(FilterPropertyObject):
+    values: RawPropertyList | ReferencePropertyValue
+
+
+RangeValue = Union[str, int, float]
+
+
+@dataclass
+class RangeObject(FilterPropertyObject):
+    gte: RangeValue | None = None
+    gt: RangeValue | None = None
+    lte: RangeValue | None = None
+    lt: RangeValue | None = None
+
+
+@dataclass
+class ParameterizedPropertyValue:
+    parameter: str
+
+
+@dataclass
+class PrefixObject(FilterPropertyObject):
+    value: str | ParameterizedPropertyValue
+
+
+@dataclass
+class ExistsObject(FilterPropertyObject):
+    ...
+
+
+@dataclass
+class ContainsAnyObject(FilterPropertyObject):
+    values: list[RawPropertyList | ReferencePropertyValue]
+
+
+@dataclass
+class ContainsAllObject(FilterPropertyObject):
+    values: list[RawPropertyList | ReferencePropertyValue]
+
+
+@dataclass
+class OverlapObject(FilterObject):
+    start_property: list[str]
+    end_property: list[str]
+    gte: RangeValue | None = None
+    gt: RangeValue | None = None
+    lte: RangeValue | None = None
+    lt: RangeValue | None = None
+
+
+Logical = Literal["and", "or", "not"]
+LOGICAL_SET = set(get_args(Logical))
+
+
+PropertyOperation = Literal[
+    "equals",
+    "in",
+    "range",
+    "prefix",
+    "exists",
+    "containsAny",
+    "containsAll",
+]
+
+ComplexOperation = Literal["matchAll", "nested", "overlaps", "hasData"]
+Operation = Union[PropertyOperation, ComplexOperation]
+
+PROPERTY_OPERATION_SET = set(get_args(PropertyOperation))
+
+FilterKey = Union[Logical, Operation]
+
+FILTER_KEY_SET = set(get_args(FilterKey))
+
+
+DSLFilter = Dict[FilterKey, Union[List, Dict]]
```

Comment:
This is gonna require some comments or simplification. I don't understand what's going on here.